### PR TITLE
feat(pacstall): Add locking mechanism

### PR DESCRIPTION
**Pull request by wizard-28**
## Approach

Add a locking mechanism to the main script, also move the argument parser to it's own function as a result.

## Checklist

- [x] I confirm that I have read the [contributing guidelines](https://github.com/pacstall/pacstall/blob/develop/CONTRIBUTING.md), and this pull request is abiding by all the clauses stated in the guideline.

### CODE REVIEW

File: src/pacstall.py [hunk 1]
```diff
@@ -59,61 +62,88 @@ def _format_action_invocation(self, action) -> str:
...
+    if len(sys.argv) == 1:
+        parser.print_help()
+        sys.exit(1)
+
+    return parser.parse_args()
+
+
+if __name__ == "__main__":
+    if getuser() == "root":
+        message.fancy("error", "Pacstall can't be run as root")
```
Comment (Henryws) [inline src/pacstall.py:134]: No contractions: `Pacstall cannot be run as root`

File: src/pacstall.py [hunk 2]
```diff
@@ -59,61 +62,88 @@ def _format_action_invocation(self, action) -> str:
...
+    if args.install or args.remove or args.upgrade:
+        global lock_file
+        lock_file = open("/var/lock/pacstall.lock", "w")
+        call("/usr/bin/sudo /usr/bin/chown root /var/lock/pacstall.lock", shell=True)
+        while True:
+            try:
+                lockf(lock_file, LOCK_EX | LOCK_NB)
+                break
+            except IOError:
+                message.fancy("error", "Another instance running lol")
```
Comment (Henryws) [inline src/pacstall.py:148]: Change it

Comment (wizard-28) [reply src/pacstall.py:148]: Oof. Right. Can you use the suggest button and suggest the change directly?

File: src/pacstall.py (hunk 1 above)
Comment (wizard-28) [reply src/pacstall.py:134]: Okay. But you should have suggested this change when I merged the PR related to this... Also use the suggest button.

File: src/pacstall.py (hunk 2 above)
Comment (Henryws) [reply src/pacstall.py:148]: Next time I will

Comment (wizard-28) [reply src/pacstall.py:148]: You know you can do it right now right?

Comment (wizard-28) [reply src/pacstall.py:148]: It appears on the ribbon when you start typing the comment here.

File: src/pacstall.py (hunk 1 above)
Comment (Henryws) [inline src/pacstall.py:134]: `message.fancy("error", "Pacstall cannot be run as root")`

File: src/pacstall.py (hunk 2 above)
Comment (Henryws) [reply src/pacstall.py:147]: `message.fancy("error", "Another instance is already running")`

**Review by Henryws — CHANGES_REQUESTED** [review]

File: src/pacstall.py (hunk 2 above)
Comment (wizard-28) [reply src/pacstall.py:147]: That's not a suggestion... Click the button (or <kbd>Ctrl</kbd> + <kbd>g</kbd>  alternatively) then write your code there.

File: src/pacstall.py (hunk 1 above)
Comment (wizard-28) [reply src/pacstall.py:134]: That's not a suggestion... Click the button (or <kbd>Ctrl</kbd> + <kbd>g</kbd>  alternatively) then write your code there.

File: src/pacstall.py [hunk 3]
```diff
@@ -59,61 +63,88 @@ def _format_action_invocation(self, action) -> str:
...
+    if len(sys.argv) == 1:
+        parser.print_help()
+        sys.exit(1)
+
+    return parser.parse_args()
+
+
+if __name__ == "__main__":
+    if getuser() == "root":
+        message.fancy("error", "Pacstall can't be run as root")
```
Comment (Henryws) [inline src/pacstall.py:136]:
```suggestion
        message.fancy("error", "Pacstall cannot be run as root")
```

File: src/pacstall.py [hunk 4]
```diff
@@ -59,61 +63,88 @@ def _format_action_invocation(self, action) -> str:
...
+
+    if args.install or args.remove or args.upgrade:
+        lock_file = open("/var/lock/pacstall.lock", "w")
+        call("/usr/bin/sudo /usr/bin/chown root /var/lock/pacstall.lock", shell=True)
+        while True:
+            try:
+                lockf(lock_file, LOCK_EX | LOCK_NB)
+                break
+            except IOError:
+                message.fancy("error", "Another instance running lol")
```
Comment (Henryws) [inline src/pacstall.py:149]:
```suggestion
                message.fancy("error", "Another instance running lol")
```

**Review by Henryws — CHANGES_REQUESTED** [review]

File: src/pacstall.py (hunk 4 above)
Comment (wizard-28) [reply src/pacstall.py:149]:
```suggestion
                message.fancy("error", "Pacstall is already running another instance")
```

How about this? This message was used in the bash version

Comment (Henryws) [reply src/pacstall.py:149]: that’s good

**Review by Henryws — APPROVED** [review]

**Review by D-Brox — APPROVED** [review]
All is right, tho I feel like the lock mechanism should be in the install and remove modules, to avoid them being run independently if we imported them in a custom python script

**Comment by wizard-28** [comment]
> All is right, tho I feel like the lock mechanism should be in the install and remove modules, to avoid them being run independently if we imported them in a custom python script

Hmm.... I guess we could always do that but can you give a situation when we'll import then elsewhere?

**Pull request merged**